packages/graphql-explorer/src/logic/QueryBuilder.ts

```diff
@@ -134,33 +138,61 @@ export default class QueryBuilder {
     }
     if (isArray(input)) {
       const arrayItems = input
-        .map((i) => this.serializeInputValue(i))
+        .map((i) => this.serializeInputValue(i, rawArgType))
         .join(', ');
       return `[${arrayItems}]`;
     }
     if (isPlainObject(input)) {
+      const objectFields = g.assertInputObjectType(rawArgType).getFields();
+
       const objectItems = Object.entries(input)
         .filter(([, v]) => v !== undefined)
-        .map(([k, v]) => `${k}: ${this.serializeInputValue(v)}`)
+        .map(
+          ([k, v]) =>
+            `${k}: ${this.serializeInputValue(v, objectFields[k].type)}`,
+        )
         .join(', ');
       return `{${objectItems}}`;
     }
-    if (input instanceof EnumValue) {
-      return input.toString();
-    }
 
     throw new Error(`invalid type for input: ${input}`);
   }
 
-  serializeArgs(args: Obj) {
+  serializeArgsInline(args: Obj, argDefinitions: g.GraphQLArgument[]) {
+    const argsByName = keyBy(argDefinitions, (a) => a.name);
     const serializedArgs = Object.entries(args)
       .filter(([, v]) => v !== undefined)
-      .map(([k, v]) => `${k}: ${this.serializeInputValue(v)}`)
+      .map(
+        ([k, v]) => `${k}: ${this.serializeInputValue(v, argsByName[k].type)}`,
+      )
       .join(', ');
 
     return serializedArgs && `(${serializedArgs})`;
   }
 
+  serializeVariableDefinitions(
+    argNames: string[],
+    argDefinitions: g.GraphQLArgument[],
+  ): [string, string] {
+    if (argNames.length === 0) {
+      return ['', ''];
+    }
+
+    const argsByName = keyBy(argDefinitions, (a) => a.name);
+
+    const definitions: string[] = [];
+    const assignments: string[] = [];
+
+    argNames.forEach((arg) => {
+      const varType = argsByName[arg].type.toString();
+      const varName = `$${arg}`;
+      definitions.push(`${varName}: ${varType}`);
+      assignments.push(`${arg}: ${varName}`);
+    });
+
+    return [`(${definitions.join(', ')})`, `(${assignments.join(', ')})`];
```

Comment:
IMO it's a better API to return an object -- you get key naming and future extendibility is easier (and typing might be easier?). Not sure I see the point in making it an array -- kind of looks like a hook but there's no obvious reason to make it that way

Reply:
yeah, more of a tuple. yes I dislike this, probably did it out of convenience. will change